src/Cli/dotnet/Commands/Tool/Restore/ToolRestoreCommand.cs

```diff
@@ -103,14 +103,24 @@ public override int Execute()
             return 0;
         }
 
+        var toolPackageRestorer = new ToolPackageRestorer(
+            _toolPackageDownloader,
+            _sources,
+            overrideSources: [],
+            _verbosity,
+            _restoreActionConfig,
+            _localToolsResolverCache,
+            _fileSystem);
+
         ToolRestoreResult[] toolRestoreResults =
             [.. packagesFromManifest
                 .AsEnumerable()
-                .Select(package => InstallPackages(package, configFile))];
+                .Select(package => toolPackageRestorer.InstallPackage(package, configFile))];
 
         Dictionary<RestoredCommandIdentifier, ToolCommand> downloaded =
-            toolRestoreResults.SelectMany(result => result.SaveToCache)
-                .ToDictionary(pair => pair.Item1, pair => pair.Item2);
+            toolRestoreResults.Select(result => result.SaveToCache)
+                .Where(item => item is not null)
+                .ToDictionary(pair => pair.Value.Item1, pair => pair.Value.Item2);
```

Comment:
Might be good to have `SaveToCache` return a tuple with named values, so you can avoid `Item1` and `Item2` usage.